cpp/src/arrow/array/diff.cc

```diff
@@ -0,0 +1,821 @@
+    return array_.IsNull(index) ? NullOr<View>() : NullOr<View>(GetView(array_, index));
+  }
+
+ private:
+  const ArrayType& array_;
+};
+
+template <typename ArrayType>
+internal::LazyRange<NullOrViewGenerator<ArrayType>> MakeNullOrViewRange(
+    const Array& array) {
+  using Generator = NullOrViewGenerator<ArrayType>;
+  return internal::LazyRange<Generator>(Generator(array), array.length());
+}
+
+/// A generic sequence difference algorithm, based on
+///
+/// E. W. Myers, "An O(ND) difference algorithm and its variations,"
+/// Algorithmica, vol. 1, no. 1-4, pp. 251–266, 1986.
+///
+/// To summarize, an edit script is computed by maintaining the furthest set of EditPoints
+/// which are reachable in a given number of edits D. This is used to compute the futhest
+/// set reachable with D+1 edits, and the process continues inductively until a complete
+/// edit script is discovered.
+///
+/// From each edit point a single deletion and insertion is made then as many shared
+/// elements as possible are skipped, recording only the endpoint of the run. This
+/// representation is minimal in the common case where the sequences differ only slightly,
+/// since most of the elements are shared between base and target and are represented
+/// implicitly.
+template <typename Iterator>
+class QuadraticSpaceMyersDiff {
+ public:
+  // represents an intermediate state in the comparison of two arrays
+  struct EditPoint {
+    Iterator base, target;
+
+    bool operator==(EditPoint other) const {
+      return base == other.base && target == other.target;
+    }
+  };
+
+  // increment the position within base (the element pointed to was deleted)
+  // then extend maximally
+  EditPoint DeleteOne(EditPoint p) const {
+    if (p.base != base_end_) {
+      ++p.base;
+    }
+    return ExtendFrom(p);
+  }
+
+  // increment the position within target (the element pointed to was inserted)
+  // then extend maximally
+  EditPoint InsertOne(EditPoint p) const {
+    if (p.target != target_end_) {
+      ++p.target;
+    }
+    return ExtendFrom(p);
+  }
+
+  // increment the position within base and target (the elements skipped in this way were
+  // present in both sequences)
+  EditPoint ExtendFrom(EditPoint p) const {
+    for (; p.base != base_end_ && p.target != target_end_; ++p.base, ++p.target) {
+      if (*p.base != *p.target) {
+        break;
+      }
+    }
+    return p;
+  }
+
+  QuadraticSpaceMyersDiff(Iterator base_begin, Iterator base_end, Iterator target_begin,
+                          Iterator target_end)
+      : base_begin_(base_begin),
+        base_end_(base_end),
+        target_begin_(target_begin),
+        target_end_(target_end),
+        endpoint_base_({ExtendFrom({base_begin_, target_begin_}).base}),
+        insert_({true}) {
+    if (std::distance(base_begin_, base_end_) ==
+            std::distance(target_begin_, target_end_) &&
+        endpoint_base_[0] == base_end_) {
+      finish_index_ = 0;
```

Comment:
Is this for the trivial case where base == target? If so, add a comment.

Reply:
will do